scripts/criu-ns

```diff
@@ -18,190 +18,195 @@ MS_SLAVE = 1 << 19
-	raise OSError(errno.EINVAL, "umount is not supported on this platform")
-else:
-	_umount.argtypes = [ctypes.c_char]
-	_umount.restype = ctypes.c_int
-
-
-def run_criu():
-	print(sys.argv)
-	os.execlp('criu', *['criu'] + sys.argv[1:])
+    _mount.argtypes = [
+        ctypes.c_char_p,
+        ctypes.c_char_p,
+        ctypes.c_char_p,
+        ctypes.c_ulong,
+        ctypes.c_void_p
+    ]
+    _mount.restype = ctypes.c_int
+
+
+def _mount_new_proc():
+    """
+    Mount new /proc filesystem.
+    """
+    if _mount(None, b"/", None, MS_SLAVE|MS_REC, None):
+        _errno = ctypes.get_errno()
+        raise OSError(_errno, errno.errorcode[_errno])
+    if _mount(b'proc', b'/proc', b'proc', 0, None):
+        _errno = ctypes.get_errno()
+        raise OSError(_errno, errno.errorcode[_errno])
+
+
+def _wait_for_process_status(criu_pid):
+    """
+    Wait for CRIU to exit and report the status back.
+    """
+    while True:
+        try:
+            (pid, status) = os.wait()
+            if pid == criu_pid:
+                return os.WEXITSTATUS(status)
+        except OSError:
+            return -251
+
+
+def run_criu(args):
+    """
+    Spawn CRIU binary
+    """
+    print(sys.argv)
+    os.execlp('criu', *['criu'] + args)
+    raise OSError(errno.ENOENT, "No such command")
 
 
 def wrap_restore():
-	# Unshare pid and mount namespaces
-	if _unshare(CLONE_NEWNS | CLONE_NEWPID) != 0:
-		_errno = ctypes.get_errno()
-		raise OSError(_errno, errno.errorcode[_errno])
-
-	criu_pid = os.fork()
-	if criu_pid == 0:
-		# Mount new /proc
-		if _mount(None, b"/", None, MS_SLAVE|MS_REC, None) != 0:
-			_errno = ctypes.get_errno()
-			raise OSError(_errno, errno.errorcode[_errno])
-
-		if _mount(b'proc', b'/proc', b'proc', 0, None) != 0:
-			_errno = ctypes.get_errno()
-			raise OSError(_errno, errno.errorcode[_errno])
-
-		# Spawn CRIU binary
-		run_criu()
-		raise OSError(errno.ENOENT, "No such command")
-
-	# Wait for CRIU to exit and report the status back
-	while True:
-		try:
-			(pid, status) = os.wait()
-			if pid == criu_pid:
-				status = os.WEXITSTATUS(status)
-				break
-		except OSError:
-			status = -251
-			break
-
-	return status
-
-
-def get_varg(args):
-	for i in range(1, len(sys.argv)):
-		if not sys.argv[i] in args:
-			continue
-
-		if i + 1 >= len(sys.argv):
-			break
+    restore_args = sys.argv[1:]
+    if '--restore-sibling' in restore_args:
+        raise OSError(errno.EINVAL, "--restore-sibling is not supported")
 
-		return (sys.argv[i + 1], i + 1)
+    # Unshare pid and mount namespaces
+    if _unshare(CLONE_NEWNS | CLONE_NEWPID) != 0:
+        _errno = ctypes.get_errno()
+        raise OSError(_errno, errno.errorcode[_errno])
 
-	return (None, None)
+    restore_detached = False
+    if '-d' in restore_args:
+        restore_detached = True
+        restore_args.remove('-d')
+    if '--restore-detached' in restore_args:
+        restore_detached = True
+        restore_args.remove('--restore-detached')
```

Comment:
A bit funny that to restore "detached" we remove "--restore-detached". =)

Reply:
Yes, it is a bit funny and paradoxical, but I hope the commit message is sufficient to explain why.